cmd/erasure.go

```diff
@@ -0,0 +1,151 @@
+	}
+	s = XLStorage{
+		disks:        make([]StorageAPI, len(disks)),
+		erasure:      erasure,
+		dataBlocks:   dataBlocks,
+		parityBlocks: parityBlocks,
+	}
+	copy(s.disks, disks)
+	return
+}
+
+// AppendWriters creates one writer for every disk of the storage array.
+// Every writer appends data to the corresponding storage disk.
+func (s *XLStorage) AppendWriters(volume, path string) []io.Writer {
+	writers := make([]io.Writer, len(s.disks))
+	for i := range writers {
+		writers[i] = StorageWriter(s.disks[i], volume, path)
+	}
+	return writers
+}
+
+// ErasureEncode encodes the given data and returns the erasure-coded data.
+// It returns an error if the erasure coding failed.
+func (s *XLStorage) ErasureEncode(data []byte) ([][]byte, error) {
+	encoded, err := s.erasure.Split(data)
+	if err != nil {
+		return nil, Errorf("failed to split data: %v", err)
+	}
+	if err = s.erasure.Encode(encoded); err != nil {
+		return nil, Errorf("failed to encode data: %v", err)
+	}
+	return encoded, nil
+}
+
+// ErasureDecode decodes the given erasure-coded data.
+// It returns an error if the decoding failed.
+func (s *XLStorage) ErasureDecode(data [][]byte) error {
+	if err := s.erasure.Reconstruct(data); err != nil {
+		return Errorf("failed to reconstruct data: %v", err)
+	}
+	return nil
+}
+
+// NewBitrotVerifier creates a new instance of a bitrot verification mechanism. The verifier implements the given algorithm.
+// It returns an error if the algorithm is not available or the key cannot be used by the algorithm.
+func NewBitrotVerifier(algorithm bitrot.Algorithm, key, checksum []byte) (*BitrotVerifier, error) {
+	if !algorithm.Available() {
+		return nil, traceError(errBitrotHashAlgoInvalid)
+	}
+	h, err := algorithm.New(key, bitrot.Verify)
+	if err != nil {
+		return nil, traceError(err)
+	}
+	return &BitrotVerifier{h, algorithm, key, checksum, false}, nil
+}
+
+// NewBitrotProtector creates a new instance of a bitrot protection mechanism. The protector implements the given algorithm.
+// It returns an error if the algorithm is not available. NewBitrotProtector will generate a new key if the algorithm requires
+// one - so the given io.Reader should return random data.
+func NewBitrotProtector(algorithm bitrot.Algorithm, random io.Reader) (key []byte, hasher bitrot.Hash, err error) {
+	if !algorithm.Available() {
+		return nil, nil, traceError(errBitrotHashAlgoInvalid)
+	}
+	if random == nil {
+		random = rand.Reader
+	}
+	key, err = algorithm.GenerateKey(random)
+	if err != nil {
+		return nil, nil, traceError(err)
+	}
+	hasher, err = algorithm.New(key, bitrot.Protect)
+	if err != nil {
+		return nil, nil, traceError(err)
+	}
+	return
+}
+
+// BitrotVerifier can be used to verify protected data.
+type BitrotVerifier struct {
+	bitrot.Hash
+
+	algorithm bitrot.Algorithm
+	key       []byte
+	sum       []byte
+	verified  bool
+}
+
+// Verify returns true iff the computed checksum of the verifier matches the the checksum provided when the verifier
```

Comment:
`iff` is the short form for `if and only if`... (a common crypto doc abbreviation)  - I can write it out if this is not clear...